### ISSUE TITLE
Update matchers page to have examples from every language

### ISSUE THREAD

**Issue by bethesque** [pact-foundation/docs.pact.io#88]
https://docs.pact.io/getting_started/matching

**Comment by YOU54F** [comment]
This looks promising as a starter for 10

https://docusaurus.io/docs/markdown-features/code-blocks#multi-language-support-code-blocks

Really digging what these guys are doing at the moment, came onto my radar the other day https://pusher.com/

![Screenshot 2022-03-17 at 19 37 37](https://user-images.githubusercontent.com/19932401/158882129-c61ccc61-8134-440d-8b1c-66d1b275ae2a.png)

**Comment by mefellows** [comment]
Yeah those code gates look handy. From memory, they also have the ability to store a language preference, which stays with you throughout the docs. This could mean we could provide a general guide that's applicable to all languages, but have language specific code blocks to accompany them.

**Comment by YOU54F** [comment]
# Pact matching features

## Matcher types

| method                      | description                                                                                                                 |
| --------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
| `boolean`                   | Match a boolean value (using equality)                                                                                      |
| `string`                    | Match a string value                                                                                                        |
| `integer`                   | Will match all numbers that are integers (both ints and longs)                                                              |
| `decimal`                   | Will match all real numbers (floating point and decimal)                                                                    |
| `hexadecimal`               | Will match all hexadecimal encoded strings                                                                                  |
| `iso8601Date`               | Will match string containing basic ISO8601 dates (e.g. 2016-01-01)                                                          |
| `iso8601DateTime`           | Will match string containing ISO 8601 formatted dates (e.g. 2015-08-06T16:53:10+01:00)                                      |
| `iso8601DateTimeWithMillis` | Will match string containing ISO 8601 formatted dates, enforcing millisecond precision (e.g. 2015-08-06T16:53:10.123+01:00) |
| `rfc3339Timestamp`          | Will match a string containing an RFC3339 formatted timestapm (e.g. Mon, 31 Oct 2016 15:21:41 -0400)                        |
| `iso8601Time`               | Will match string containing times (e.g. T22:44:30.652Z)                                                                    |
| `ipv4Address`               | Will match string containing IP4 formatted address                                                                          |
| `ipv6Address`               | Will match string containing IP6 formatted address                                                                          |
| `uuid`                      | Will match strings containing UUIDs                                                                                         |
| `email`                     | Will match strings containing Email address                                                                                 |


## Matcher Links

- Pact Concepts
  - https://docs.pact.io/getting_started/matching
- Ruby 
  - https://docs.pact.io/implementation_guides/ruby/matching
- Java
  - https://docs.pact.io/implementation_guides/jvm/matching
- Javascript
  - https://docs.pact.io/implementation_guides/javascript/readme#matching
- Rust
  - https://docs.pact.io/implementation_guides/rust/pact_matching
- PHP
  - https://github.com/pact-foundation/pact-php/#create-mock-response
- Go
  - https://docs.pact.io/implementation_guides/go/readme#matching
- .net
  - https://github.com/pact-foundation/pact-net/blob/master/docs/messaging-pacts.md
- C++
  - https://docs.pact.io/implementation_guides/cpp/consumer#matchers
- Python
  - https://github.com/pact-foundation/pact-python/#expecting-variable-content
- Swift
  - https://github.com/DiUS/pact-consumer-swift#matching
  - https://github.com/surpher/PactSwift/wiki/Matchers
- Scala
  - http://io.itv.com/scala-pact/advanced/matching-rules.html
  - https://github.com/jbwheatley/pact4s



## Regular Expressions

Ruby
```ruby title="regex"
animal_service.given("an alligator named Mary exists").
  upon_receiving("a request for an alligator").
  with(
    method: "get",
    path: "/alligators/Mary",
    headers: {"Accept" => "application/json"}).
  will_respond_with(
    status: 200,
    headers: {"Content-Type" => "application/json"},
    body: {
      name: "Mary",
      dateOfBirth: Pact.term(
        generate: "02/11/2013",
        matcher: /\d{2}\/\d{2}\/\d{4}/)
    })
```

Java
```java title='regex'
service {
    uponReceiving('a request')
    withAttributes(method: 'get', path: '/')
    withBody {
      name(~/\w+/, 'harry')
      surname regexp(~/\w+/, 'larry')
      position regexp(~/staff|contractor/, 'staff')
      happy(true)
    }
}
```

Go
```go title='regex'
pact.
    AddInteraction().
    Given("User sally exists").
    UponReceiving("A request to login with user 'sally'").
    WithRequest(request{
      Method: "GET",
      Path:   term("/users/10", "/users/[0-9]+"),
    }).
    WillRespondWith(dsl.Response{
      Status:  200,
      Body:    dsl.Match(model.User{}),
      Headers: commonHeaders,
    })
```

JS
```js title='regex'
const { term } = pact
provider.addInteraction({
  state: "Has some animals",
  uponReceiving: "a request for an animal",
  withRequest: {
    method: "GET",
    path: "/animals/1",
  },
  willRespondWith: {
    status: 200,
    headers: {
      "Content-Type": "application/json; charset=utf-8",
    },
    body: {
      id: 100,
      name: "billy",
      gender: term({
        matcher: "F|M",
        generate: "F",
      }),
    },
  },
})
```

C++
```cpp title='regex'
this.pact
    .UponReceiving("a request to retrieve all events")
        .WithRequest(HttpMethod.Get, "/events")
        .WithHeader("Accept", "application/json")
    .WillRespond()
        .WithStatus(HttpStatusCode.OK)
        .WithHeader("Content-Type", "application/json; charset=utf-8")
        .WithJsonBody(Match.MinType(new
        {
            eventId = Match.Type("3E83A96B-2A0C-49B1-9959-26DF23F83AEB"),
            timestamp = Match.Type("2014-06-30T01:38:00.8518952"),
            eventType = Match.Regex("SearchView", "SearchView|DetailsView")
        }, 1));
```


python

```python title='regex'
from pact import Term
...
body = {
    'username': 'UserA',
    'last_modified': Term('\d+-\d+-\d+T\d+:\d+:\d+', '2016-12-15T20:16:01')
}

(pact
 .given('UserA exists and is not an administrator')
 .upon_receiving('a request for UserA')
 .with_request('get', '/users/UserA/info')
 .will_respond_with(200, body=body))
```

**Comment by YOU54F** [comment]
Ruby regex backslash escaping causing js pain again

https://pact-foundation.slack.com/archives/C9VBGLUM9/p1651866363050989
